content/doc/developer/testing/index.adoc

```diff
@@ -110,7 +110,30 @@ Most Java IDEs should be able to run JUnit tests and report on the results.
 == What to Test
 Now that we can write a basic test, we should discuss what you should be testing…
 
-TODO: Unit testing of your code, as much as possible. JenkinsRule testing: creating jobs that use your build steps, running them, asserting on the output
+The test pyramid is a concept used in software development to describe the ideal distribution of different types of automated tests that make up a test suite.
+The idea was introduced by Mike Cohn and is visualized as a pyramid to help illustrate the optimal proportion of each type of test.
+Here's a breakdown of the three main layers from the bottom to the top of the pyramid:
+
+1. *Unit Tests*: At the base of the pyramid, unit tests are the most numerous.
+These tests are focused on checking the smallest parts of an application, such as individual functions or methods.
+Unit tests are quick to execute and help ensure that each component of the software works as expected in isolation.
+
+2. *Integration Tests*: The middle layer of the pyramid consists of integration tests.
+These are fewer in number compared to unit tests.
+Integration tests verify that different parts of the system work together as intended.
+For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] to test the integration of the plugin into Jenkins.
+This can be a test for a new build step and testing this with a test pipeline.
+
+3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
+These tests simulate real user scenarios from start to finish.
+They interact with the application as a user would, testing the complete flow of the system.
+For Jenkins, integration tests including the UI, make use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call, which is used to create a call to an endpoint to then assert something on the result.
+The End-To-End Tests are done via the link:https://github.com/jenkinsci/acceptance-test-harness[`acceptance-test-harness`] or can be done with a real browser, as can be seen in the link:https://github.com/jenkinsci/warnings-ng-plugin/tree/main/ui-tests[warnings-ng ui-tests].
```

Comment:
I would move the HTML unit tests to topic 2 ( but maybe I am too pedantic as I am teaching Mike Cohns definition in my testing courses 🙈)

```suggestion
For Jenkins, integration tests including the UI make use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call, which is used to create a call to an HTML-Unit endpoint to then assert something on the result. 

This can be a test for a new build step and testing the results in the UI with a test pipeline.

3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
These tests simulate real user scenarios from start to finish.
They interact with the application as a user would, testing the complete flow of the system.
The End-To-End Tests are done with a real browser in combination with the link:https://github.com/jenkinsci/acceptance-test-harness[`acceptance-test-harness`], as can be seen in the link:https://github.com/jenkinsci/warnings-ng-plugin/tree/main/ui-tests[warnings-ng ui-tests].

These tests are typically not used in Jenkins plugins because they are too complex to create and maintain.
```